src/v/utils/bottomless_token_bucket.h

```diff
@@ -0,0 +1,150 @@
+    using tokens_t = int64_t; // [tokens]
+    using clock = ss::lowres_clock;
+    using time_res_t = std::chrono::milliseconds;
+
+    // Algoritms rely on time_res_t to be subsecond
+    static_assert(time_res_t::period::num == 1);
+
+    // Maximum supported width of a bucket ~ 25 days
+    static constexpr time_res_t max_width{std::numeric_limits<int32_t>::max()};
+
+    // To avoid dynamic checking of acceptable width and quota maximums,
+    // quota is limited in the way to avoid integer overflows in any condition.
+    // This effectively maxes it to ~ 2 Ttokens/s. If that is ever found not
+    // enough, this maximum should be raised to full int64_t, and both
+    // set_quota() and set_width() should dynamically check that the calculated
+    // burst does not overflow.
+    static constexpr quota_t max_quota{
+      static_cast<quota_t>(std::numeric_limits<int32_t>::max())
+      * time_res_t::period::den};
+
+    bottomless_token_bucket(
+      const quota_t quota, const time_res_t width) noexcept {
+        // do boundary checks on params and calculate _burst
+        set_quota_impl(quota);
+        set_width_impl(width);
+        update_burst_tokens();
+    }
+
+    void set_quota(const quota_t quota) noexcept {
+        set_quota_impl(quota);
+        update_burst_tokens();
+    }
+
+    /// @post 0 < return value <= max_quota
+    quota_t quota() const noexcept { return _quota; }
+
+    void set_width(const time_res_t width) noexcept {
+        set_width_impl(width);
+        update_burst_tokens();
+    }
+
+    /// Record usage of tokens
+    /// @param n Number of tokens used
+    /// @param now Time to record at
+    void use(const tokens_t n, const clock::time_point now) noexcept {
+        vassert(n >= 0, "Cannot use negative number of tokens");
+        refill(now);
+        _tokens -= n;
+    }
+
+    /// Current level of tokens
+    tokens_t tokens() const noexcept { return _tokens; }
+
+private:
+    void set_quota_impl(const quota_t quota) noexcept {
+        vassert(quota > 0, "Token bucket quota must be positive ({})", quota);
+        vassert(
+          quota <= max_quota,
+          "Token bucket quota too large ({} > {})",
+          quota,
+          max_quota);
+        _quota = quota;
+    }
+
+    void set_width_impl(const time_res_t width) noexcept {
+        vassert(
+          width.count() > 0,
+          "Token bucket width must be positive ({})",
+          width.count());
+        vassert(
+          width <= max_width,
+          "Token bucket width too large ({} > {})",
+          width.count(),
+          max_width.count());
+        _width = width;
+    }
+
+    void update_burst_tokens() noexcept {
+        _burst = muldiv(_quota, _width.count(), time_res_t::period::den);
+        _tokens = std::min(_burst, _tokens);
+    }
+
+    void refill(const clock::time_point now) noexcept {
+        const auto delta = std::chrono::duration_cast<time_res_t>(
+          now - _last_check);
```

Comment:
what is the expected behavior here for the first invocation? since this is a lowres_clock (steady_clock) the normal usage of `now` is not wall-clock time like it is with a system_lock, but rather time since boot-up. so it not guaranteed that now-last_check (where last_check is the zero point default initialized) is necessarily going to be a large duration.

Reply:
`_last_check` will be zero. You're right that `now()` will not necessarily be big enough for `delta` to saturate the token bucket, but practical effect of that will be limitied to faster response of the bucket to the negative side. Token bucket is an integrator, so any specific initial state (as well as a one time externally induced change to the state) only cause short lived perturbation. That's why I have decided to avoid any nontrivial loging involving "initial" state.

On the other hand, even that can probably be mitigated by initializing `_last_check` to `-max_width`. Let me see if that can be feasible.

Reply:
thanks for the explanation and makes sense. whatever you decide to do here sounds reasonable, but i'd add a comment in the implementation explaining it. once i discovered that lowres clock is steady clock, i end up double checking lowres math everything i see it, so calling it out makes sense.

Reply:
Improved that, now token buket is always initialized full.